src/llama.cpp

```diff
@@ -18436,282 +18517,185 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
+            if (ab_map.find(name) == ab_map.end()) {
+                ab_map[name] = lora_weight(nullptr, cur);
+            } else {
+                ab_map[name].b = cur;
             }
-            base_t = ggml_dup_tensor(lora_ctx, ml->get_tensor_meta(base_name.c_str()));
         } else {
-            base_t = ggml_dup_tensor(lora_ctx, model_t);
-        }
-        ggml_set_name(base_t, base_name.c_str());
-
-        // allocate in backend buffer
-        ggml_backend_buffer_t lora_buf = ggml_backend_alloc_ctx_tensors_from_buft(lora_ctx, ggml_backend_cpu_buffer_type());
-        if (lora_buf == nullptr) {
-            LLAMA_LOG_ERROR("%s: error: failed to allocate lora tensors\n", __func__);
-            return 1;
+            LLAMA_LOG_WARN("%s: discard tensor '%s'\n", __func__, cur->name);
         }
+    }
 
-        // load tensor data
-        auto load_tensor = [&read_buf, &fin](const tensor_meta & tensor_meta, ggml_tensor * tensor) {
-            read_buf.resize(ggml_nbytes(tensor));
-            fin.seek(tensor_meta.offset, SEEK_SET);
-            fin.read_raw(read_buf.data(), ggml_nbytes(tensor));
-            ggml_backend_tensor_set(tensor, read_buf.data(), 0, read_buf.size());
-        };
-        load_tensor(metaA, loraA);
-        load_tensor(metaB, loraB);
-
-        // load base model tensor data
-        if (ml) {
-            ml->load_data_for(base_t);
+    // add tensors
+    for (auto & it : ab_map) {
+        std::string name = it.first;
+        const char * cname = name.c_str();
+        lora_weight & w = it.second;
+        GGML_ASSERT(w.a != nullptr);
+        GGML_ASSERT(w.b != nullptr);
+        int il = -1;
+        sscanf(cname, "blk.%d.", &il);
+        struct ggml_context * dev_ctx; // device ctx
+        if (il >= 0) {
+            dev_ctx = ctx_map.at(model->buft_layer[il].buft);
+        } else if (strstr(cname, "tok") == 0) {
+            dev_ctx = ctx_map.at(model->buft_input.buft);
+        } else if (strstr(cname, "output") == 0) {
+            dev_ctx = ctx_map.at(model->buft_output.buft);
         } else {
-            ggml_backend_tensor_copy(model_t, base_t);
-        }
-
-        if (ggml_is_quantized(base_t->type) && !warned) {
-            LLAMA_LOG_WARN("%s: warning: using a lora adapter with a quantized model may result in poor quality, "
-                            "use a f16 or f32 base model with --lora-base\n", __func__);
-            warned = true;
-        }
-
-        if (base_t->ne[0] != loraA->ne[1] || base_t->ne[1] != loraB->ne[1]) {
-            LLAMA_LOG_ERROR("%s: incompatible tensor dimensions (%" PRId64 " and %" PRId64 ");"
-                            " are you sure that this adapter is for this model?\n", __func__, base_t->ne[0], loraA->ne[1]);
-            ggml_free(lora_ctx);
-            ggml_backend_buffer_free(lora_buf);
-            ggml_backend_free(backend_cpu);
-            return 1;
+            LLAMA_LOG_WARN("%s: discard tensor '%s'\n", __func__, cname);
```

Comment:
I don't think it is a good idea to ignore extra tensors, it likely means that the lora is not valid.

Reply:
Right, we already had a check before this line to ignore `optimizer.*` tensors, so here it's not necessary. This line will now throw an exception: https://github.com/ggerganov/llama.cpp/pull/8332/commits/79e2982788b0102aabb098b1a3d6227a7e32a483